lib/matplotlib/tests/test_colorbar.py

```diff
@@ -235,6 +235,23 @@ def test_remove_from_figure(use_gridspec):
     assert (pre_position.get_points() == post_position.get_points()).all()
 
 
+def test_remove_from_figure_cl():
+    """
+    Test `remove` with constrained_layout
+    """
+    fig, ax = plt.subplots(constrained_layout=True)
+    sc = ax.scatter([1, 2], [3, 4], cmap="spring")
+    sc.set_array(np.array([5, 6]))
+    fig.draw_without_rendering()
+    pre_position = ax.get_position()
+    cb = fig.colorbar(sc)
+    cb.remove()
+    fig.draw_without_rendering()
+    post_position = ax.get_position()
+    np.testing.assert_allclose(pre_position.get_points(),
+                               post_position.get_points())
+
```

Comment:
Possibly add an assert that cb.ax isn't in the `ax._colorbars` list since that is the addition you're adding above?

Reply:
My understanding is that we don't usually check private attributes?

Reply:
In general not. Private attributes cannot guarantee any user-facing functionality. They cannot be the primary check for a feature. But if we think that testing internal state helps to catch bugs, we're free to test those as well. IMHO one could argue that here.

Reply:
If `cb.ax` is still in `ax._colorbars` this test will raise with the `AttributeError: 'NoneType' object has no attribute 'transSubfigure'` of the original issue.  Thats what this test checks, I'm not convinced we need to test the implementation.